Allow range types to be used for enrich matching

mjmbischoff opened the pull request:
Currently we hardcode that the data type for the match field on the enrich index, when the MATCH policy is used, this is (always) keyword. This breaks look ups when source indicies use a [rangetype](https://www.elastic.co/guide/en/elasticsearch/reference/current/range.html) for lookup. For example when a ip_range is used the literal value from the source is indexed as a keyword and therefor doesn't match any of the ip's.

This PR improves the code by checking for the field type on the source indicies and then proceeds to use the range type in the enrich index if this is the type for all source indicies. If the types are different across source indices, we fallback to keyword. For non-range types we also fallback to keyword, emulating the current behavior. 

---

- [x] Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- [x] Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md)?
- [x] If submitting code, have you built your formula locally prior to submission with `gradle check`?
- [x] If submitting code, is your pull request against master? Unless there is a good reason otherwise, we prefer pull requests against master and will backport as needed.
- [x] If submitting code, have you checked that your submission is for an [OS and architecture that we support](https://www.elastic.co/support/matrix#show_os)?

mjmbischoff committed:
Allowing enrich to MATCH range types

mjmbischoff committed:
Allowing enrich to MATCH range types - fixing unchecked cast

mjmbischoff committed:
Allowing enrich to MATCH range types - fixing checkstyle rule violations

mjmbischoff committed:
Allowing enrich to MATCH range types - running spotless

jbaiera commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -232,19 +238,76 @@ private static void validateField(Map<?, ?> properties, String fieldName, boolea
+            case "date_range":
+            case "ip_range":
+                return type;
+            case "text": // exact matching only ?
```
I think this should be exact matching only for sure.

mjmbischoff replied:
Think this a moot point, now as if we did support fancier text matching it would be a different policy. Removed case

jbaiera commented on x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunnerTests.java:
```diff
@@ -255,6 +255,93 @@ public void testRunnerGeoMatchType() throws Exception {
+        IndexResponse indexRequest = client().index(
+            new IndexRequest().index(sourceIndexName)
+                .id("id")
+                .source("{" + "\"subnet\":" + "\"10.0.0.0/8\"," + "\"department\":\"research\"" + "}", XContentType.JSON)
```
Can we test the scenario where `subnet` and `department` are fields underneath an object field (like `organization.subnet` and `organization.department`)?

mjmbischoff replied:
Added additional tests, did reveal an issue with the code. Good one.

jbaiera reviewed (commented):
The code looks pretty good here, though I can't seem to remember if we had plans to separate ranges out from the `match` enrich policy type. Searching for data using indexed ranges feels closer to how `geo_match` works, but the underlying query DSL is indeed closer to how `match` would write it.

@martijnvg do you remember if that was the plan or not?

martijnvg commented:
Apologies for responding late to this PR.
Like James, I think the code in this PR looks good.

> @martijnvg do you remember if that was the plan or not?

Yes, the plan was to introduce a new enrich policy type for ranges.
This has the added benefit that actual range can be queried in addition to literal values.
Przemko started with a pr for this a while ago: #65781 (but this got stalled due to inactivity).
However maybe this enhancement to the match type can exist next with the future range policy type?
What do you think?

jbaiera commented:
Since we have two separate policy types (match and geo_match) for different query types, I think I'd lean more toward keeping design parity and introducing a third policy type. While the underlying query DSL still uses a match query, we always intended the match policy type to enrich based on singular values. Adding a range policy type gives us the flexibility to support future growth, while adding range support in the match policy type may muddy the waters going forward should the functionality be duplicated. 

Does the policy type pertain more to the query DSL created, or is it more concerned with the field type of the enrich key? I'll throw the team discuss label on here to see if we can reach consensus.

elasticmachine commented:
Pinging @elastic/es-core-features (Team:Core/Features)

martijnvg commented:
> Since we have two separate policy types (match and geo_match) for different query types, I think I'd lean more toward keeping design parity and introducing a third policy type. While the underlying query DSL still uses a match query, we always intended the match policy type to enrich based on singular values. Adding a range policy type gives us the flexibility to support future growth, while adding range support in the match policy type may muddy the waters going forward should the functionality be duplicated.

Muddying the waters when it comes to policy types is a valid concern if this change would get merged. Thinking from that perspective, maybe the match processor should not be changed to use more field types for the match field.

> Does the policy type pertain more to the query DSL created, or is it more concerned with the field type of the enrich key?

I think it is both. It controls what field type to use the match field (during policy execution) and the query that is executed against this field (in the processor implementation).

mjmbischoff committed:
Splitting MATCH_POLICY into MATCH_POLICY / RANGE_POLICY
Adding tests for multi level + fix.

mjmbischoff commented:
> Apologies for responding late to this PR.

No worries, you should enjoy your PTO.

> Przemko started with a pr for this a while ago: #65781 (but this got stalled due to inactivity).

Missed this, will have a closer look at it to see what I've missed.

Added the changes to make it a separate policy but realized I forgot about the configuring part. I'll see if I can pick this up later.

mjmbischoff committed:
spotless fix

mjmbischoff committed:
fixing test

mjmbischoff committed:
spotless again

mjmbischoff committed:
Merge branch 'elastic:master' into range_enrich

mjmbischoff commented:
Looking into https://github.com/elastic/elasticsearch/pull/65781 , it seems a format field is introduced for date but no such parameter is described for https://www.elastic.co/guide/en/elasticsearch/reference/7.14/query-dsl-term-query.html in our docs (which seems incomplete as it doesn't describe the range matching either) it is described in the type description for the mapping https://www.elastic.co/guide/en/elasticsearch/reference/7.14/range.html

It is described for the range query but that is for taking a range as the input to match on a value (or possible range), while I'm sure some user can find a reason to match like that I think that would be a different policy (and PR) it does make me wonder if the naming for the policy we introduce here is the right one. And potentially re-evaluate the split? 

I should add some tests for dates including exotic date formats that it won't magically pick to check if we need the format configured and if it needs to be user specified or copied from the mapping of the source / enrich index.

martijnvg commented:
@mjmbischoff Looks like the `format` option in #65781 is used for `format` option of the date field in the enrich index. However perhaps this isn't needed and can the `format` option be extracted from the mapping of the source index during enrich policy execution.

> It is described for the range query but that is for taking a range as the input to match on a value (or possible range), while I'm sure some user can find a reason to match like that I think that would be a different policy (and PR) it does make me wonder if the naming for the policy we introduce here is the right one. And potentially re-evaluate the split?

Right, perhaps ranges should be handled by a different kind of policy, like the one proposed in #65781.

mjmbischoff committed:
Added tests and fix for date_range format

mjmbischoff commented:
Going with RANGE_TYPE, allowing  RANGE_MATCH_TYPE or better name to be used for matching ranges against values, ranges.

Added tests for date_range which indeed triggered on the format being missing. Similar to types, requiring a singular distinct format specification or otherwise we fail and inform the user. If someone feels they need can argue that they need multiple source indicies with different format fields feel free to open an ER arguing your case. Merging these is non-trivial, because one format might mask the other etc.

mjmbischoff committed:
spotless

mjmbischoff committed:
Merge branch 'elastic:master' into range_enrich

mjmbischoff commented:
@martijnvg @jbaiera think it's ready.

martijnvg commented:
Removed team-discuss label, as there is agreement that a dedicated policy type is right approach for supporting ranges (which is the direction that this PR takes)

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -232,19 +238,134 @@ private static void validateField(Map<?, ?> properties, String fieldName, boolea
+                    // date_range types mappings allow for the format to be specified, should be preserved in the created index
+                    if (matchType.equals("date_range")) {
+                        List<String> formatEntries = collapseFieldMappingsValue(response, toFielMappingFormat(matchField));
+                        if (types.isEmpty()) {
```
Shouldn't this check ``formatEntries.isEmpty()` instead?

mjmbischoff replied:
Good catch, addressed in https://github.com/elastic/elasticsearch/pull/76110/commits/910db97ce4ddf0dace3c81415d9ceabc42bd3062

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -232,19 +238,134 @@ private static void validateField(Map<?, ?> properties, String fieldName, boolea
+            policy.getIndices().stream().toArray(String[]::new)
+        ).fields(policy.getMatchField());
+
+        client.execute(GetFieldMappingsAction.INSTANCE, fieldMappingsRequest, listener.delegateFailure((l, response) -> {
```
Instead of invoking the get field mapping action, can range type (with a potential date format) be resolved from the response of the get index request that happened a few steps earlier?

mjmbischoff replied:
good catch, I changed it so it reuses the index request see https://github.com/elastic/elasticsearch/pull/76110/commits/910db97ce4ddf0dace3c81415d9ceabc42bd3062

Interesting how the java object structure is different different and, of course, that we add logic at all the call sites instead of some centralized util or convenience methods on the response object itself.  Passing this on to future us.

martijnvg replied:
iirc the get index api returns the metadata as is stored in cluster state. Whereis get mappings api has a different response structure for consuming the response in Kibana.

martijnvg reviewed (commented):
Thanks! I left a few comments.

mjmbischoff committed:
Fixing review comments
- checking wrong collection
- removing mapping request and reusing earlier index request

mjmbischoff committed:
Merge branch 'elastic:master' into range_enrich

mjmbischoff committed:
pulling in https://github.com/elastic/elasticsearch/pull/65781/files#diff-be6652cff75cf1e283570c14fd5dea042ddb586ac5645f7deef6ef3716732906 from probakowski

mjmbischoff committed:
pulling in https://github.com/elastic/elasticsearch/pull/65781/files#diff-d893db3430284d2635214c41f0cb66b3ca154d6cd73c29c5c97031ece40d2904 from probakowski
+ spotless

mjmbischoff commented on x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunnerTests.java:
```diff
@@ -1143,6 +1356,311 @@ public void testRunnerTwoObjectLevelsSourceMapping() throws Exception {
+        // Validate Index is read only
+        ensureEnrichIndexIsReadOnly(createdEnrichIndex);
+    }
+
```
Transplated tests, small adaptations but could mostly run unmodified see https://github.com/elastic/elasticsearch/pull/76110/commits/8dc44eb07fc8d4391c0c2b95d654d86f128fbe66 and https://github.com/elastic/elasticsearch/pull/76110/commits/fc81fdfc0eb99b495add9eed14feca0652c58e45

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -232,19 +235,136 @@ private static void validateField(Map<?, ?> properties, String fieldName, boolea
+                }
+            }
+            result = property(result, parts[parts.length - 1]);
+            return String.valueOf(result.getOrDefault(property, "unknown"));
```
I don't see handling for `unknown` as return value. Is this something that should be done? Or maybe just `null` should be returned?

mjmbischoff replied:
I think your right, I  did a pass to clean/simply things but am now no longer checking for unknown and for isEmpty instead. IT does still help with more human friendly error messages. Going to revisit this.

mjmbischoff replied:
adressed in https://github.com/elastic/elasticsearch/pull/76110/commits/44cab8e54cb59099114a07af78626c1130a74d09 / https://github.com/elastic/elasticsearch/pull/76110/commits/b6e71e8db37c15de85ebeb3173f82c0e023f447e

mjmbischoff committed:
fixing strict checking

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -232,19 +235,136 @@ private static void validateField(Map<?, ?> properties, String fieldName, boolea
+                            (builder) -> builder.field("type", matchType).field("doc_values", false),
+                            resultListener
+                        );
+                    } else if (formatEntries.size() > 1) {
```
I think the `format` option in field mappers can only just have a single value. It should be possible that a date range mapping has more than 1 date format strings. (note that multiple date formats can be specified inside a single value using the `||` notation)

So I don't think this needs to be checked?

mjmbischoff replied:
You can have multiple source indicies, therefor I assumed there can be distinct mappings that might conflict, is there something else stopping this?

> indices
> (Required, String or array of strings) One or more source indices used to create the enrich index.
Honestly I haven't seen it get used in the field, but we seem to allow it.

martijnvg replied:
I see, I haven't thought about that. I think it is for now best to fail with an error if that were to happen. 

Although this code could also combine the different formats (from different date_range mappings from different indices) into a single date format using `||` notation. But let's keep it as is know, this could always be changed later.

mjmbischoff replied:
Check, I also had [this](https://github.com/elastic/elasticsearch/pull/76110#issuecomment-907542635) comment a little higher on this issue.

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -232,19 +235,136 @@ private static void validateField(Map<?, ?> properties, String fieldName, boolea
 
+    private void resolveRangeTypeEnrichMapping(
+        EnrichPolicy policy,
+        GetIndexResponse response,
```
Maybe pass down here a `Map<String, Object>` instead of a `GetIndexResponse`? Since that the only thing from the response used here.

martijnvg replied:
And then maybe replace `collapseFieldMappingsValue(...)` and `property(...)` method with `ObjectPath#eval(...)`?

mjmbischoff replied:
`ObjectPath#eval(...)` was definitely the thing I was looking for :sweat_smile: 

Passing around the map makes it a bit unclear whats in the map / structure of it, I don't know if that sounds weird?

martijnvg replied:
> Passing around the map makes it a bit unclear whats in the map / structure of it, I don't know if that sounds weird?

I think if the parameter is named `mappingSource` or something like that, that it isn't an issue.

mjmbischoff replied:
Addressed in https://github.com/elastic/elasticsearch/pull/76110/commits/44cab8e54cb59099114a07af78626c1130a74d09

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -232,19 +235,136 @@ private static void validateField(Map<?, ?> properties, String fieldName, boolea
+    ) {
+        String matchField = policy.getMatchField();
+        List<String> types = collapseFieldMappingsValue(response, toFieldMappingProperty(matchField, "type"));
+        if (types.isEmpty()) {
```
I think with the above simplification (using `ObjectPath#eval(...)`), this can just be a null check? (instead of checking for empty and more than size 1 (which shouldn't be possible?))

mjmbischoff replied:
Similar to below, a user is currently allowed to specify multiple source indicies which might have conflicting mappings.

martijnvg reviewed (commented):
Looks good @mjmbischoff! I left a few more comments around simplification.

mjmbischoff committed:
Fixing test

mjmbischoff committed:
Dammit, use refactor.

mjmbischoff committed:
Fixing tests

mjmbischoff committed:
Cleaning up IndexResponse handling

mjmbischoff committed:
Eval returns null, filter null values
Fixing bad replace.

martijnvg commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into range_enrich

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -232,19 +243,104 @@ private static void validateField(Map<?, ?> properties, String fieldName, boolea
+
+        Set<String> types = matchFieldMappings.stream().map(map -> map.get("type")).collect(Collectors.toSet());
+
+        if (types.isEmpty()) {
```
This is more of a style comment, but maybe let's do the error case in the else clause? I think it is easier to read, first the two valid cases and then everything else is an error case. Like this:

```java
if (types.isEmpty()) {
   // first valid case
} else if (types.size() == 1) {
   // second valid case
} else {
   // error case
}
```

mjmbischoff replied:
Considered the same thing 😅 but wanted to keep the delta small. I'll switch it around, also fiddled with a Optional but didn't seem to improve things, Kinda missing an Object to represent nothing/singular/multiple options.

mjmbischoff replied:
addressed in https://github.com/elastic/elasticsearch/pull/76110/commits/2924d2bb76c3cd027a1bf129f2de39b08a2461aa

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -232,19 +243,104 @@ private static void validateField(Map<?, ?> properties, String fieldName, boolea
+                case "date_range":
+                    Set<String> formatEntries = matchFieldMappings.stream().map(map -> map.get("format")).collect(Collectors.toSet());
+
+                    if (formatEntries.isEmpty()) {
```
Same here (first valid cases and then error case)?

mjmbischoff replied:
addressed in https://github.com/elastic/elasticsearch/pull/76110/commits/114a452ca9ea4fd844788e53406c0b69da75415d

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -294,15 +390,17 @@ private void prepareAndCreateEnrichIndex() {
             // This disables eager global ordinals loading for all fields:
             .put("index.warmer.enabled", false)
             .build();
-        CreateIndexRequest createEnrichIndexRequest = new CreateIndexRequest(enrichIndexName, enrichIndexSettings);
```
I think this can be restored? I think this `resolveEnrichMapping(...)` and `resolveRangeTypeEnrichMapping(...)` can now just return the mapping via the return value? (since `resolveRangeTypeEnrichMapping(...)` doesn't invoke the get field mapping api any more).

mjmbischoff replied:
addressed in https://github.com/elastic/elasticsearch/pull/76110/commits/2924d2bb76c3cd027a1bf129f2de39b08a2461aa

martijnvg reviewed (commented):
Thanks for the quick update! I left a few more comments.

mjmbischoff committed:
switching clauses around to get happy flows on top.

mjmbischoff committed:
removing callback
adding randomness to test

mjmbischoff committed:
prettier exception description
omit format field if default

mjmbischoff commented on x-pack/plugin/enrich/qa/common/src/main/java/org/elasticsearch/test/enrich/CommonEnrichRestTestCase.java:
```diff
@@ -179,7 +250,13 @@ public static String createSourceIndexMapping() {
             + "\"tldRank\":{\"type\":\"keyword\"},"
-            + "\"tld\":{\"type\":\"keyword\"}"
+            + "\"tld\":{\"type\":\"keyword\"},"
+            + "\"date\":{\"type\":\"date_range\", \"format\": \"yyyy-MM-dd\"},"
```
addressed in https://github.com/elastic/elasticsearch/pull/76110/commits/2924d2bb76c3cd027a1bf129f2de39b08a2461aa

martijnvg reviewed (approved):
Thanks for iterating on this PR @mjmbischoff.
I think this PR looks good now.
@jbaiera do you think this PR is good to get merged?

jbaiera reviewed (approved):
LGTM, thanks for the changes!

mjmbischoff commented:
@martijnvg I did the honors, is it ok if I leave backporting (if applicable) to you?

martijnvg commented:
> I did the honors

🎉 

> is it ok if I leave backporting (if applicable) to you?

Yes, I will do the back porting to the 7.x branch.
